gateway/src/shard/processor.rs

```diff
@@ -253,27 +279,32 @@ impl ShardProcessor {
             };
 
             match msg {
-                Message::Binary(bytes) => {
-                    trace!("Payload: {}", String::from_utf8_lossy(&bytes));
-
-                    match serde_json::from_slice(&bytes) {
-                        Ok(event) => break event,
-                        Err(why) => {
-                            warn!("Error deserializing {:?}: {:?}", bytes, why);
-                        },
-                    }
+                Message::Binary(bin) => {
+                    self.inflater.extend(&bin[..]);
+
+                    let msg_or_error =
+                        match self.inflater.msg().map_err(|source| Error::Decompressing {
```

Comment:
This is pretty ugly, let's move out this `?` to another binding and then match on that